### PR TITLE
Various cleanups and enhancements

### PR DESCRIPTION
**Pull request by jkuebart**
Only a single PR since several changes have been submitted at once.

I created the »base« branch in order to separate previously reviewed changes. Going forward, it would be useful to retire branches once their PRs and reviews are finished.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -116,54 +89,45 @@ int Main(vector<string> args)
 
 				// push input parameters on stack
 				lua_newtable(L);
-					lua_pushfield(L, "rule", rule.c_str());
-					lua_pushfield(L, "matched", sv.c_str(), sv.length());
-					lua_pushfield(L, "line", sv.line_info().first);
-					lua_pushfield(L, "column", sv.line_info().second);
-					lua_pushfield(L, "choice", sv.choice());
+					lua_pushfield(L, "rule",    rule.c_str());
```
Comment (jkuebart) [inline src/main.cpp]: I'm not a big fan of this type of alignment: As be seen here and in the change to `peglib.h`, it leads to gratuitous changes on unrelated lines. Therefore, I would suggest to use only single spaces.

File: src/luautils.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+#pragma once
+
+#include <string>
+#include "lua/src/lua.hpp"
+
+using namespace std;
+
+// utilities for lua stack manipulation
+struct LuaStackPtr{
+	LuaStackPtr(const int idx_):idx(idx_){}
```
Comment (jkuebart) [inline src/luautils.h]: Declare `explicit` to prevent unwanted automatic conversions to go unnoticed.

File: src/luautils.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	const int idx;
+};
+
+struct StringPtr{
+	StringPtr(const char* c_, const size_t len_):c(c_),len(len_){}
```
[the remark this answers is not in the document]
Comment (jkuebart) [reply src/luautils.h]: I think a constructor taking a `std::string` would be beneficial for this class:
```c++
explicit StringPtr(const std::string& s)
: c(s.data())
, len(s.size())
{}
```

File: src/main.cpp (hunk 1 above)
Comment (jkuebart) [reply src/main.cpp]: `lua_pushfield(L, "rule", rule)` avoids re-measuring the string's length.

File: src/main.cpp [hunk 4]
```diff
@@ -174,19 +138,27 @@ int Main(vector<string> args)
 		
 		parser[r.c_str()].enter = [r](const char* s, size_t n, any& dt) {
 			auto& indent = *dt.get<int*>();
-			for(int i=0; i<indent; i++){cout << "|  ";}
-			string beginning(s, n);
-			if(beginning.length()>10){beginning = beginning.substr(0,10) + "...";}
-			for(auto& c:beginning){if(c == '\n'){c = '\\';}}
-			cout << r << " => \"" << beginning << "\"?" << endl;
+			cout << repeat("|  ", indent) << r << " => \"" << shorten(string(s, n), 40) << "\"?" << endl;
```
Comment (jkuebart) [inline src/main.cpp]: nit: `shorten(string(s, std::min(38, n)), 40)` may avoid an unnecessary copy of a long string.

File: src/main.cpp [hunk 5]
```diff
@@ -174,19 +138,27 @@ int Main(vector<string> args)
...
+		parser[r.c_str()].leave = [r, &L](const char* s, size_t n, size_t matchlen, any& value, any& dt) {
 			auto& indent = *dt.get<int*>();
 			indent--;
-			for(int i=0; i<indent; i++){cout << "|  ";}
-			cout << "`-> " << (match ? "match" : "failed") << endl;
+			cout << repeat("|  ", indent) << "`-> ";
+			if(success(matchlen)){
+
+				// display "match", the matched string and the result of the reduction
+				cout << "match: \"" << shorten(string(s, matchlen), 40) << "\" -> ";
```
Comment (jkuebart) [inline src/main.cpp]: If `shorten` is always called with the pattern `shorten(string(s, len), maxlen)`, it might make sense to change the signature to `shorten(const char* s, std::size_t len, std::size_t maxlen)` to avoid gratuitous string copies.

File: src/main.cpp [hunk 6]
```diff
@@ -174,19 +138,27 @@ int Main(vector<string> args)
...
+				lua_getglobal(L, "stringify");
+				lua_push(L, value.get<LuaStackPtr>());
+				lua_pcall(L, 1, 1, 0);
+				string output = lua_tostring(L, -1);
+				cout << shorten(output, 40) << endl;
```
Comment (jkuebart) [inline src/main.cpp]: this can still be used with the above suggestion as `shorten(output.data(), output.size(), 40)`.

File: src/main.cpp [hunk 7]
```diff
@@ -222,3 +195,4 @@ int main(int c, char** v){
 #endif
 	return result;
 }
+
```
Comment (jkuebart) [inline src/main.cpp]: spurious line

File: src/stringutils.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
+
+#pragma once
+
+#include <iostream>
```
Comment (jkuebart) [inline src/stringutils.h]: `ostream` is sufficient.

File: src/stringutils.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+#include <string>
+
+// helper for repeating output
+struct repeat{
+	repeat(const char* s_, size_t num_):s(s_), num(num_){}
+	const char* s;
+	size_t num;
+};
+ostream& operator<< (ostream& stream, const repeat rep){
```
Comment (jkuebart) [inline src/stringutils.h]: `const repeat&` is more efficient.

Comment (jkuebart) [reply src/stringutils.h]: use `std::ostream`

File: src/stringutils.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	for(int i=0; i<rep.num; ++i){
+		stream << rep.s;
+	}
+	return stream;
+}
+
+// shorten string and remove line breaks and tabs
+string shorten(const string& txt, const size_t numchars){
```
Comment (jkuebart) [inline src/stringutils.h]: use `std::string` here and below.

File: test/calculator.lua [hunk 11]
```diff
@@ -61,25 +61,12 @@ function MulOp(params)
...
 
 function Num(params)
-	print("                          ", dump(params))
 	return params.matched
 end
 
 function default(params)
-	print("                          ", dump(params))
 	return params.matched
 end
```
Comment (jkuebart) [inline test/calculator.lua]: two spurious lines before EOF

File: src/luautils.h [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+template<typename TKey>
+void lua_opensubtable(lua_State *L, const TKey key){
+	lua_push(L, key);
+	lua_newtable(L);
+}
+
+auto lua_loadutils(lua_State *L){
+	const auto utils = R"(
```
Comment (jkuebart) [inline src/luautils.h]: indentation looks off

**Comment by jkuebart** [comment]
These commits also contain several whitespace problems, such as indent with spaces in `peglib.h`, trailing whitespace in `main.cpp` and `peglib.h` and trailing newlines in `calculator.lua`, `luautils.h` and `main.cpp`

File: src/main.cpp [hunk 13]
```diff
@@ -64,30 +30,37 @@ int Main(vector<string> args)
...
 	// load text to parse
 	ifstream textfile {args[2]};
-	string text { istreambuf_iterator<char>(textfile), istreambuf_iterator<char>() };
-	if ("" == text){
-		cout << "error loading \"" << args[2] << "\": file not found or empty" << endl;
+	if(textfile.fail()){
+		cerr << "error loading \"" << args[2] << "\": file not found" << endl;
 		return EXIT_FAILURE;
-	}		
+    }
```
Comment (jkuebart) [inline src/main.cpp]: indentation looks off

File: src/luautils.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (mqnc) [reply src/luautils.h]: Main purpose of this class is to hand c-strings to the overloaded lua_push function as single arguments. As there is also a lua_push that takes std::strings, the StringPtr would not be needed there. Do y... Would there still be a use for the suggested constructor? To make StringPtr more general maybe?

[the remark this answers is not in the document]
Comment (jkuebart) [reply src/luautils.h]: It was intended for the call `lua_pushfield(L, rule)` in `main.cpp` line 92, but by the time I got there I realised it could just pass the `std::string`. So *you're* right, it would actually be confusing to have a `std::string` c'tor and this `struct` should stay as it is ;-)
